src/jit/lsraxarch.cpp

```diff
@@ -2497,32 +2497,73 @@ void LinearScan::TreeNodeInfoInitSIMD(GenTreeSIMD* simdTree, TreeNodeInfo* info)
 void LinearScan::TreeNodeInfoInitHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, TreeNodeInfo* info)
 {
     NamedIntrinsic intrinsicID = intrinsicTree->gtHWIntrinsicId;
-    InstructionSet isa         = compiler->isaOfHWIntrinsic(intrinsicID);
+    InstructionSet isa         = Compiler::isaOfHWIntrinsic(intrinsicID);
     if (isa == InstructionSet_AVX || isa == InstructionSet_AVX2)
     {
         SetContainsAVXFlags(true, 32);
     }
-    info->srcCount += GetOperandInfo(intrinsicTree->gtOp.gtOp1);
-    if (intrinsicTree->gtGetOp2IfPresent() != nullptr)
+    GenTree* op1   = intrinsicTree->gtOp.gtOp1;
+    GenTree* op2   = intrinsicTree->gtOp.gtOp2;
+    info->srcCount = 0;
+
+    if (op1 != nullptr)
     {
-        info->srcCount += GetOperandInfo(intrinsicTree->gtOp.gtOp2);
+        if (op1->OperIsList())
+        {
+            for (GenTreeArgList* list = op1->AsArgList(); list != nullptr; list = list->Rest())
+            {
+                info->srcCount += GetOperandInfo(list->Current());
+            }
+        }
+        else
+        {
+            info->srcCount += GetOperandInfo(op1);
+        }
     }
 
-#ifdef _TARGET_X86_
-    if (intrinsicTree->gtHWIntrinsicId == NI_SSE42_Crc32)
+    if (op2 != nullptr)
     {
-        // CRC32 may operate over "byte" but on x86 only RBM_BYTE_REGS can be used as byte registers.
-        //
-        // TODO - currently we use the BaseType to bring the type of the second argument
-        // to the code generator. May encode the overload info in other way.
-        var_types srcType = intrinsicTree->gtSIMDBaseType;
-        if (varTypeIsByte(srcType))
+        info->srcCount += GetOperandInfo(op2);
+    }
+
+    switch (intrinsicID)
+    {
+        case NI_SSE41_BlendVariable:
         {
-            LocationInfoListNode* op2Info = useList.GetSecond(INDEBUG(intrinsicTree->gtGetOp2()));
-            op2Info->info.setSrcCandidates(this, RBM_BYTE_REGS);
+            if (!compiler->canUseVexEncoding())
+            {
+                // SSE4.1 blendv* hardcode the mask vector (op3) in XMM0
+                LocationInfoListNode* op3Info = useList.Begin()->Next()->Next();
+                LocationInfoListNode* op2Info = useList.Begin()->Next();
+                op2Info->info.isDelayFree     = true;
+                op3Info->info.isDelayFree     = true;
+                info->internalFloatCount      = 1;
+                info->setInternalCandidates(this, RBM_XMM0);
+                info->isInternalRegDelayFree = true;
+                info->hasDelayFreeSrc        = true;
```

Comment:
@mikedn I changed to emit a `movaps xmm0, op3's reg` in codegen and set XMM0 as the internal candidate. Now, the code generation looks correct. Thanks a lot!
```csharp
var vf1 = Unsafe.Read<Vector128<float>>(ptr1);
var vf2 = Unsafe.Read<Vector128<float>>(ptr2);
var vf3 = Unsafe.Read<Vector128<float>>(ptr3);
var vf4 = Sse.Add(vf1, vf2);
var vf5 = Sse41.BlendVariable(vf1, vf3, vf2);
var vf6 = Sse41.BlendVariable(vf1, vf5, vf4);
var vf7 = Sse41.BlendVariable(vf1, vf2, vf6);
```
```asm
N0053: 0001E1 movups   xmm1, xmmword ptr [rax]
IN0054: 0001E4 movups   xmm2, xmmword ptr [rdi+16]
IN0055: 0001E8 movaps   xmm3, xmm6
IN0056: 0001EB addps    xmm3, xmm1
IN0057: 0001EE movaps   xmm0, xmm1
IN0058: 0001F1 movaps   xmm4, xmm6
IN0059: 0001F4 blendvps  xmm4, xmm2
IN005a: 0001F9 movaps   xmm0, xmm3
IN005b: 0001FC movaps   xmm2, xmm6
IN005c: 0001FF blendvps  xmm2, xmm4
IN005d: 000204 movaps   xmm0, xmm2
IN005e: 000207 blendvps  xmm6, xmm1
```

Reply:
Can we log a bug to track figuring out why we can't directly set `op3` to get `xmm0` assigned?

Reply:
> Can we log a bug to track figuring out why we can't directly set op3 to get xmm0 assigned?

There is no bug.

Reply:
> There is no bug.

What do you mean? It seems like a bug that we have to explicitly `movaps xmm0, op3Reg` rather than having the register allocator do it for us...

Reply:
I agree - it would be good to understand why you can't set the candidates on op3 to make this happen.

Reply:
>  It seems like a bug that we have to explicitly movaps xmm0, op3Reg 

It's not a bug, that's how it works. There are plenty of other places in codegen that do this. You know that it's supposed to be in XMM0 so you can do it yourself. Having the RA do it means inserting an extra `GT_COPY` node.

Reply:
@CarolEidt I think the real question here is how come out of 3 cases the RA never managed to allocate `xmm0` to the def position so that a copy isn't necessary at the use position. It's either simply how things work today or perhaps there's an issue with this weird "binary-node-that's-really-ternary-and-has-null-op2" thing.

Reply:
Ah, stupid me. Of course it can't generate something like
```asm
IN005c: 0001FF blendvps  xmm0, xmm4, xmm0
IN005e: 000207 blendvps  xmm6, xmm1, xmm0
```
due to op3 being delay free.

Reply:
> It's not a bug, that's how it works. There are plenty of other places in codegen that do this. You know that it's supposed to be in XMM0 so you can do it yourself. Having the RA do it means inserting an extra GT_COPY node.

The RA won't insert a `GT_COPY` for this case, because it is a single register, so the contract between the RA and Codegen is that, if the value is in a different register, codegen is responsible for doing the move. However, this also provides a hint that it should be assigned that register if possible. Is there some other factor here that I've missed?

Reply:
> However, this also provides a hint that it should be assigned that register if possible. 

Where? At def or at use?

Reply:
> Where? At def or at use?

At both, if there is no conflict. In this case, since the context is the use, it would otherwise be on the use. It will also generate a fixed use on the register at the use location.

Reply:
It may be useful to look at a slightly simpler example:
```C#
        var vf1 = Unsafe.Read<Vector128<float>>(ptr1);
        var vf2 = Unsafe.Read<Vector128<float>>(ptr2);
        var vf4 = Sse.Add(vf2, vf1);
        var vf3 = Unsafe.Read<Vector128<float>>(ptr3);
        var vf5 = Sse41.BlendVariable(vf2, vf3, vf4);
```
generates
```asm
       0F1002               movups   xmm0, xmmword ptr [rdx]
       410F1008             movups   xmm1, xmmword ptr [r8]
       0F28D1               movaps   xmm2, xmm1
       0F58D0               addps    xmm2, xmm0
       410F1019             movups   xmm3, xmmword ptr [r9]
       0F28C2               movaps   xmm0, xmm2
       660F3814CB           blendvps  xmm1, xmm3
```
while the ideal asm would be more like
```asm
       0F1002               movups   xmm0, xmmword ptr [rdx]
       410F1008             movups   xmm1, xmmword ptr [r8]
       0F58D0               addps    xmm0, xmm1
       410F1019             movups   xmm3, xmmword ptr [r9]
       660F3814CB           blendvps  xmm1, xmm3
```
It's not clear why `addps`'s destination is `xmm2` rather than `xmm0`.

Reply:
> It's not clear why addps's destination is xmm2 rather than xmm0.

Because I did not set the register-candidate of op3 to XMM0, and set XMM) as the candidate of the internal register.

Reply:
> Because I did not set the register-candidate of op3 to XMM0, and set XMM0 as the candidate of the internal register.

Huh?

Reply:
Setting the op3's candidate to XMM0 works fine with the simple example, but does not work with the complex example.
So I set the XMM0 as the candidate of the internal register to try to hold XMM0 for `blendv*`.

Reply:
> So I set the XMM0 as the candidate of the internal register to try to hold XMM0 for blendv*.

@mikedn I am not sure if this is correct.

Reply:
@mikedn Can we directly insert "copy op3reg to XMM0" in codgen without holding XMM0 as an internal register?

Reply:
>  I am not sure if this is correct.

It seems correct but it's kind of backwards.

> Can we directly insert "copy op3reg to XMM0" in codgen without holding XMM0 as an internal register?

This works fine for me:
```asm
LocationInfoListNode* op2Info = useList.Begin()->Next();
op2Info->info.isDelayFree     = true;
LocationInfoListNode* op3Info = op2Info->Next();
op3Info->info.isDelayFree     = true;
op3Info->info.setSrcCandidates(this, RBM_XMM0);
info->isInternalRegDelayFree = true;
info->hasDelayFreeSrc        = true;
```

Reply:
Do we need `info->isInternalRegDelayFree = true;` here?